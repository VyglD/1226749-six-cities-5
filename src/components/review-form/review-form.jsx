import React from "react";

const MIN_CHARACTERS = 50;
const MAX_CHARACTERS = 300;

class ReviewForm extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      rating: 0,
      review: ``,
      isValid: false,
    };

    this.handleFieldChange = this.handleFieldChange.bind(this);
    this.handleRatingChange = this.handleRatingChange.bind(this);
    this.handleSubmitComment = this.handleSubmitComment.bind(this);
  }

  checkFormValidity() {
    const {review, rating} = this.state;

    if (review.length > MIN_CHARACTERS
      && review.length < MAX_CHARACTERS
      && rating > 0) {
      this.setState({isValid: true});
    } else {
      this.setState({isValid: false});
    }
  }

  handleRatingChange(evt) {
    this.setState(
        {rating: parseInt(evt.target.value, 10)},
        this.checkFormValidity
    );
  }

  handleFieldChange(evt) {
    const {name, value} = evt.target;
    this.setState(
        {[name]: value},
        this.checkFormValidity
    );
  }

  handleSubmitComment(evt) {
    evt.preventDefault();

    // вызов метода для изменения массива комментов
  }

  render() {
    return (
      <form
        className="reviews__form form"
        action="#"
        method="post"
        onSubmit={this.handleSubmitComment}
      >
        <label className="reviews__label form__label" htmlFor="review">Your review</label>
        <div className="reviews__rating-form form__rating">
          <input
            className="form__rating-input visually-hidden"
            name="rating"
            value="5"
            id="stars-5"
            type="radio"
            onChange={this.handleRatingChange}
          />
          <label htmlFor="stars-5" className="reviews__rating-label form__rating-label" title="perfect">
            <svg className="form__star-image" width="37" height="33">
              <use xlinkHref="#icon-star"></use>
            </svg>
          </label>

          <input
            className="form__rating-input visually-hidden"
            name="rating"
            value="4"
            id="stars-4"
            type="radio"
            onChange={this.handleRatingChange}
          />
          <label htmlFor="stars-4" className="reviews__rating-label form__rating-label" title="good">
            <svg className="form__star-image" width="37" height="33">
              <use xlinkHref="#icon-star"></use>
            </svg>
          </label>

          <input
            className="form__rating-input visually-hidden"
            name="rating"
            value="3"
            id="stars-3"
            type="radio"
            onChange={this.handleRatingChange}
          />
          <label htmlFor="stars-3" className="reviews__rating-label form__rating-label" title="not bad">
            <svg className="form__star-image" width="37" height="33">
              <use xlinkHref="#icon-star"></use>
            </svg>
          </label>

          <input
            className="form__rating-input visually-hidden"
            name="rating"
            value="2"
            id="stars-2"
            type="radio"
            onChange={this.handleRatingChange}
          />
          <label htmlFor="stars-2" className="reviews__rating-label form__rating-label" title="badly">
            <svg className="form__star-image" width="37" height="33">
              <use xlinkHref="#icon-star"></use>
            </svg>
          </label>

          <input
            className="form__rating-input visually-hidden"
            name="rating"
            value="1"
            id="stars-1"
            type="radio"
            onChange={this.handleRatingChange}
          />
          <label htmlFor="stars-1" className="reviews__rating-label form__rating-label" title="terribly">
            <svg className="form__star-image" width="37" height="33">
              <use xlinkHref="#icon-star"></use>
            </svg>
          </label>
        </div>
        <textarea
          className="reviews__textarea form__textarea"
          id="review"
          name="review"
          onChange={this.handleFieldChange}
          placeholder="Tell how was your stay, what you like and what can be improved">
        </textarea>
        <div className="reviews__button-wrapper">
          <p className="reviews__help">
            To submit review please make sure to set <span className="reviews__star">rating</span> and describe your stay with at least <b className="reviews__text-amount">50 characters</b>.
          </p>
          <button
            className="reviews__submit form__submit button"
            type="submit"
            disabled={this.state.isValid ? `` : `disabled`}
          >
            Submit
          </button>
        </div>
      </form>
    );
  }
}

export default ReviewForm;
