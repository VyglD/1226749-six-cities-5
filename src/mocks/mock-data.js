const Key = {
  ENTER: `Enter`,
  ESC: `Escape`,
  SPACE: ` `,
  TAB: `Tab`,
  SHIFT: `Shift`,
  UP: `ArrowUp`,
  DOWN: `ArrowDown`,
};

const CITIES = [
  `Paris`,
  `Cologne`,
  `Brussels`,
  `Amsterdam`,
  `Hamburg`,
  `Dusseldorf`
];

const city = CITIES[3];
const activeCity = CITIES[0];

const HousingType = {
  APARTMENT: `Apartment`,
  ROOM: `Private Room`,
  HOUSE: `House`,
  HOTEL: `Hotel`
};

const SortType = {
  DEFAULT: {
    value: `Popular`,
    method: () => {},
  },
  PRICE_TO_HIGH: {
    value: `Price: low to high`,
    method: (i, j) => (i.cost - j.cost),
  },
  PRICE_TO_LOW: {
    value: `Price: high to low`,
    method: (i, j) => (j.cost - i.cost),
  },
  RATE: {
    value: `Top rated first`,
    method: (i, j) => (j.rate - i.rate),
  },
};

const mockOffers = [
  {
    id: `1`,
    city: CITIES[0],
    latitude: 48.95,
    longitude: 2.35,
    title: `Title of first server offer`,
    photos: [`img/photo1`, `img/photo11`],
    description: `Description of first server offer`,
    isPremium: true,
    housingType: HousingType.APARTMENT,
    rate: 4.5,
    rooms: 4,
    guests: 4,
    cost: 111,
    features: [`Washer`, `Towels`, `Fridge`],
    owner: {
      avatar: `img/avatar-angelina.jpg`,
      name: `Anna`,
      isSuper: true,
    },
    cityInfo: {
      name: CITIES[0],
      latitude: 48.85661,
      longitude: 2.351499,
      zoom: 12,
    }
  },
  {
    id: `2`,
    city: CITIES[0],
    latitude: 48.87,
    longitude: 2.33,
    title: `Title of cheapest offer`,
    photos: [],
    description: `Description of cheapest offer`,
    isPremium: false,
    housingType: HousingType.HOTEL,
    rate: 3,
    rooms: 3,
    guests: 3,
    cost: 1,
    features: [],
    owner: {
      avatar: `img/avatar-angelina.jpg`,
      name: `Angelina`,
      isSuper: true,
    },
    cityInfo: {
      name: CITIES[0],
      latitude: 48.85661,
      longitude: 2.351499,
      zoom: 12,
    }
  },
  {
    id: `3`,
    city: CITIES[1],
    latitude: 50.95,
    longitude: 6.95,
    title: `Title of most expensive offer`,
    photos: [`img/photo3`],
    description: `Description of most expensive offer`,
    isPremium: true,
    housingType: HousingType.APARTMENT,
    rate: 4.5,
    rooms: 4,
    guests: 4,
    cost: 4444,
    features: [`Washer`, `Towels`, `Fridge`],
    owner: {
      avatar: `img/avatar-angelina.jpg`,
      name: `Anna`,
      isSuper: true,
    },
    cityInfo: {
      name: CITIES[1],
      latitude: 50.938361,
      longitude: 6.959974,
      zoom: 13,
    },
  },
  {
    id: `4`,
    city: CITIES[2],
    latitude: 50.83,
    longitude: 4.34,
    title: `Title of top rated offer`,
    photos: [`img/photo4`, `img/photo44`],
    description: `Description of top rated offer`,
    isPremium: true,
    housingType: HousingType.HOUSE,
    rate: 5,
    rooms: 1,
    guests: 1,
    cost: 222,
    features: [`Washer`, `Towels`, `Fridge`],
    owner: {
      avatar: `img/avatar-angelina.jpg`,
      name: `Anna`,
      isSuper: true,
    },
    cityInfo: {
      name: CITIES[1],
      latitude: 50.846557,
      longitude: 4.351697,
      zoom: 13,
    },
  },
];

const mockOffersFavoriteIds = [`1`, `2`];

const mockReview = {
  id: `1`,
  name: `Bob`,
  photo: `img/photo`,
  rate: 3,
  date: new Date(0),
  text: `comment`,
};

const mockFunction = () => {};

const mockEvent = {
  preventDefault: mockFunction,
};

const mockEscKeyEvent = Object.assign(
    {},
    mockEvent,
    {
      key: Key.ESC,
    }
);

export {
  Key,
  city,
  activeCity,
  CITIES,
  SortType,
  mockOffers,
  mockOffersFavoriteIds,
  mockReview,
  mockFunction,
  mockEvent,
  mockEscKeyEvent,
};
