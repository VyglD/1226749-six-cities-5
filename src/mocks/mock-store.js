import configureStore from 'redux-mock-store';
import {
  mockOffers,
  mockOffersFavoriteIds,
  activeCity,
} from "./mock-data";

const createStore = configureStore();

const mockStore = createStore({
  OFFERS: {
    allOffers: mockOffers,
  },
  FAVORITES: {
    favoriteIds: mockOffersFavoriteIds,
  },
  OFFER: {
    openedOfferId: null,
    reviews: [],
    nearOffers: [],
  },
  USER: {
    isLogin: false,
    email: ``,
    avatar: ``,
    isSuper: false,
  },
  REDIRECT: {},
  CITY: {
    activeCity,
  },
});

export {mockStore};
