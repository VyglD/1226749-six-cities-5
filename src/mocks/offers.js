import {HOUSING_TYPE} from "../const";

const AVATAR_URL = `https://api.adorable.io/avatars/74`;
const BIG_NUMBER = 999;

const MAX_PHOTOS = 5;
const MAX_RATE = 5;
const MAX_ROOMS = 3;
const MAX_GUESTS = 4;

const features = [
  `Wi-Fi`,
  `Heating`,
  `Kitchen`,
  `Fridge`,
  `Washing machine`,
  `Coffee machine`,
  `Dishwasher`,
  `Towels`,
  `Baby seat`,
  `Cabel TV`
];

const getRandomInteger = (a = 0, b = 1) => {
  const lower = Math.ceil(Math.min(a, b));
  const upper = Math.floor(Math.max(a, b));

  return Math.floor(lower + Math.random() * (upper - lower + 1));
};

const generatePhotos = () => {
  return new Array(getRandomInteger(MAX_PHOTOS))
    .fill()
    .map(() => {
      return `https://picsum.photos/id/${getRandomInteger(BIG_NUMBER)}/260/200`;
    });
};

const getRandomElement = (array) => {
  return array[getRandomInteger(0, array.length - 1)];
};

const shuffleArray = (arr) => {
  return arr.slice().sort(() => {
    return 0.5 - Math.random();
  });
};

const getRandomSubArray = (array) => {
  const length = array.length;
  const half = Math.floor(length / 2);
  const start = getRandomInteger(0, half);
  const end = getRandomInteger(half, length);

  return shuffleArray(array).slice(start, end);
};

export default [
  {
    id: `1`,
    city: `Amsterdam`,
    title: `Beautiful & luxurious apartment at great location`,
    photos: [`/img/apartment-01.jpg`, ...generatePhotos()],
    description: [
      `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
      `An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.`
    ],
    isPremium: true,
    housingType: HOUSING_TYPE.APARTMENT,
    rate: getRandomInteger(MAX_RATE),
    rooms: getRandomInteger(MAX_ROOMS),
    guests: getRandomInteger(MAX_GUESTS),
    cost: 120,
    features: getRandomSubArray(features),
    owner: {
      avatar: `/img/avatar-angelina.jpg`,
      name: `Angelina`,
      isSuper: true
    }
  },
  {
    id: `2`,
    city: `Amsterdam`,
    title: `Wood and stone place`,
    photos: [`/img/room.jpg`, ...generatePhotos()],
    description: [
      `This average-sized, rectangular bedroom has matching wooden furniture.
        The floor is stone and the walls are painted and decorated with a wallpaper border.
        Light is provided by table lamps and a ceiling light.
        The room is done in colors that remind you of Easter and overall has a quirky look.
        Among the first things one notices walking in is a collection of action figures.`
    ],
    isPremium: false,
    housingType: HOUSING_TYPE.ROOM,
    rate: getRandomInteger(MAX_RATE),
    rooms: getRandomInteger(MAX_ROOMS),
    guests: getRandomInteger(MAX_GUESTS),
    cost: 80,
    features: getRandomSubArray(features),
    owner: {
      avatar: `${AVATAR_URL}/${Math.random()}`,
      name: `Mark`,
      isSuper: false
    }
  },
  {
    id: `3`,
    city: `Amsterdam`,
    title: `Canal View Prinsengracht`,
    photos: [`/img/apartment-02.jpg`, ...generatePhotos()],
    description: [
      `This spacious, square office has matching metal and plastic furniture.
        The floor is carpeted and the walls are painted with a paneled dado.
        Light is provided by wall lamps and a ceiling light.`,
      `The room is done in a vintage theme in warm colors and overall looks like an old castle.
        Among the first things one notices walking in is art supplies on the desk.`
    ],
    isPremium: false,
    housingType: getRandomElement(Object.values(HOUSING_TYPE)),
    rate: getRandomInteger(MAX_RATE),
    rooms: getRandomInteger(MAX_ROOMS),
    guests: getRandomInteger(MAX_GUESTS),
    cost: 132,
    features: getRandomSubArray(features),
    owner: {
      avatar: `${AVATAR_URL}/${Math.random()}`,
      name: `Julia`,
      isSuper: false
    }
  },
  {
    id: `4`,
    city: `Amsterdam`,
    title: `Nice, cozy, warm big bed apartment`,
    photos: [`/img/apartment-03.jpg`, ...generatePhotos()],
    description: [
      `This average-sized, square dining room has matching wooden furniture.`,
      `The seating is cushioned.`,
      `The floor is tiled and the walls are painted and decorated with a wallpaper border.`,
      `Light is provided by floor lamps and a ceiling light.`,
      `The room is done in colors that remind you of a coffee shop and overall looks very modern.
        Among the first things one notices walking in is a strange clock.`
    ],
    isPremium: true,
    housingType: getRandomElement(Object.values(HOUSING_TYPE)),
    rate: getRandomInteger(MAX_RATE),
    rooms: getRandomInteger(MAX_ROOMS),
    guests: getRandomInteger(MAX_GUESTS),
    cost: 180,
    features: getRandomSubArray(features),
    owner: {
      avatar: `${AVATAR_URL}/${Math.random()}`,
      name: `Tom`,
      isSuper: true
    }
  }
];
