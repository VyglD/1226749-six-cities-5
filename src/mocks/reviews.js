export default [
  {
    offerId: `1`,
    name: `Max`,
    rate: 4,
    date: `April 2019`,
    text: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.
      The building is green and from 18th century.`
  },
  {
    offerId: `2`,
    name: `Bob`,
    rate: 4,
    date: `May 2017`,
    text: `Comment #2`
  },
  {
    offerId: `2`,
    name: `Scarlett`,
    rate: 2,
    date: `May 2018`,
    text: `Comment #3`
  },
  {
    offerId: `3`,
    name: `Bruce`,
    rate: 3,
    date: `June 2020`,
    text: `Comment #4`
  },
  {
    offerId: `3`,
    name: `Amanda`,
    rate: 4,
    date: `January 2018`,
    text: `Comment #5`
  },
  {
    offerId: `3`,
    name: `Maria`,
    rate: 5,
    date: `March 2011`,
    text: `Comment #6`
  }
];
