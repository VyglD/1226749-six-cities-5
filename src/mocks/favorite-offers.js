import offers, {getRandomSubArray} from "./offers";

const MAX_FAVORITE_OFFERS = 3;

export default getRandomSubArray(offers)
  .slice(0, MAX_FAVORITE_OFFERS)
  .map((offer) => offer.id);
