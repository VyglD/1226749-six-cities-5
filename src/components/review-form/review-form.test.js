import React from "react";
import renderer from "react-test-renderer";
import {ReviewForm} from "./review-form";
import {mockOffers, mockFunction} from "../../mocks/mock-data";

it(`Render correctly component ReviewForm`, () => {
  const component = renderer.create(
      <ReviewForm
        chosenOffer={mockOffers[0]}
        postNewReview={mockFunction}
      />,
      {
        createNodeMock: () => {
          return {};
        }
      }
  ).toJSON();

  expect(component).toMatchSnapshot();
});
