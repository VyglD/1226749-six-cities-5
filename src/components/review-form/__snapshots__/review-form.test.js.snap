// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render correctly component ReviewForm 1`] = `
<form
  action="#"
  className="reviews__form form"
  method="post"
  onSubmit={[Function]}
>
  <label
    className="reviews__label form__label"
    htmlFor="review"
  >
    Your review
  </label>
  <div
    className="reviews__rating-form form__rating"
  >
    <input
      className="form__rating-input visually-hidden"
      id="stars-5"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={5}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="stars-5"
      title="perfect"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      className="form__rating-input visually-hidden"
      id="stars-4"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={4}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="stars-4"
      title="perfect"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      className="form__rating-input visually-hidden"
      id="stars-3"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={3}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="stars-3"
      title="perfect"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      className="form__rating-input visually-hidden"
      id="stars-2"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={2}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="stars-2"
      title="perfect"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      className="form__rating-input visually-hidden"
      id="stars-1"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={1}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="stars-1"
      title="perfect"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
  </div>
  <textarea
    className="reviews__textarea form__textarea"
    id="review"
    name="review"
    onChange={[Function]}
    placeholder="Tell how was your stay, what you like and what can be improved"
  />
  <div
    className="reviews__button-wrapper"
  >
    <p
      className="reviews__help"
    >
      To submit review please make sure to set 
      <span
        className="reviews__star"
      >
        rating
      </span>
       and describe your stay with at least 
      <b
        className="reviews__text-amount"
      >
        50 characters
      </b>
      .
    </p>
    <button
      className="reviews__submit form__submit button"
      onClick={[Function]}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;
